### ISSUE TITLE
Feature Request: Keep Bundled Dependencies

### ISSUE THREAD

**Issue by ike18t** [ng-packagr/ng-packagr#1432]
It would be great if bundled dependencies would not get deprecated.

The company I work for has a large monolithic Rails + Angular application. It used to be developed and maintained by a single team but is now being developed and maintained by several teams. We've started adding new pages/sections/components by taking a micro-frontend approach. New pages are standalone npm projects with their own git repo, build pipeline, npm package, etc. The original application pulls in each micro-frontend project by specifying the latest version in its package.json file and lazy loads them in by the base path. We share a common version of angular (and angular dependencies) via peer dependencies but the version of other libraries and dependencies can differ between micro-frontend packages (or differ with the original monolithic application). Bundled dependencies provides the ability to have dependencies differ. A recent example is that we had one package using date-fns v1 and another using date-fns v2.

The TL;DR is that having the ability to bundle the dependencies allows us to prevent version conflicts between our micro-frontend applications and integrate more easily into the host application.

**Comment by antontemchenko** [comment]
Hi, like @ike18t we are actively using bundled dependencies.
We build angular modules as libraries.
Then those modules are asynchronously loaded into different applications.
"Bundled dependencies" feature allows us to pack all required dependencies with library, if application doesn't provide them.
"Bundled dependencies" allows us to reach multiple goals:
- to separately update and deploy libraries (as they are loaded asynchronously into applications)
- to use separate versions of third party libraries if we stop maintaining one of modules. (we can update third party library in app and bundle older version with non-maintained module)

**Comment by tgrux** [comment]
+1  
As the above users, we are using NPM to construct a Micro UI architecture that relies on bundled dependencies for UMD package creation.  Bundled dependencies with shadow DOM allow us to provide freedom to those micro applications.

**Comment by flash-me** [comment]
Gosh, why I'm not stumbled over this earlier....

**To start with the suggestion of @alan-agius4 to use `bundledDependencies` (see [here](https://github.com/ng-packagr/ng-packagr/issues/1690#issuecomment-661128236)):**

This is **not** working. What I did was to add an entry to `package.json` like this:
```
"bundledDependencies": [
    "is-glob"
  ]
```
This is recognised correctly and the information is there, as we can see here:
![image](https://user-images.githubusercontent.com/5240450/106733598-62534180-6612-11eb-87dd-3f0b97976fa0.png)

But the function provided for the `external` property returns `true`:
![image](https://user-images.githubusercontent.com/5240450/106733902-b6f6bc80-6612-11eb-87ba-5b6c9ebee32f.png)

**_Given that, `rollup` will still treat that module as an external and thus won't bundle it._**

@alan-agius4 Since you suggested this approach, but this doesn't work, I think there is a bug?

cheers
flash ⚡

**Comment by flash-me** [comment]
B2T:

This issue is not only related to a "micro frontends" approach, it's a general one. We will encounter this situation every time where two parts are depending on different major version of third party lib. Let's say your `app` uses `lib` version 2, but you also use a library that itself depends on `lib` version 1. What now? 🤷🏽‍♂️

> As the above users, we are using NPM to construct a Micro UI architecture that relies on bundled dependencies for UMD package creation. Bundled dependencies with shadow DOM allow us to provide freedom to those micro applications.

I'm really confused at this point. Shadow DOM only scopes DOM and CSS, so I have no idea what you are talking about? 
The appropriate way for doing this would be realms. [But this is still a stage 2 proposal](https://github.com/tc39/proposal-realms).
Anyway

### Let's talk about solutions that work.

@ike18t @antontemchenko @tgrux there are many approaches to accomplish this. The best one would be direct support of ng-packagr, but that's not the case. (But who knows the future?)
The approach I chosed for this had the following aspects in mind:
- As few impact on the workflow as possible
- Easy to adapt
- No additional dependencies

Let's have a look what we have. The goal is to create angular libraries with some dependencies are already bundled with it.
Since this is the @ng-packagr repo and you folks probably build your libs using it, we can make use of the toolings that  `ng-packagr` itself is making use of and thus avoid installing additional stuff with **one exception (optional)**.

Concretely speaking, we will use @rollup and some plugins. The solutions is just to "rebundle" the outputs of `ng-packagr` and include the dependencies you are interested in. In this example all of the tools are implicitly installed, except `rollup-plugin-terser`, which I manually installed afterwards. (It's optional, maybe you don't want to minify, or want to directly use terser..) 

Here is the snipped:
```js
import nodeResolve from '@rollup/plugin-node-resolve';
import commonJs from '@rollup/plugin-commonjs';
import sourceMaps from 'rollup-plugin-sourcemaps';
import {terser} from 'rollup-plugin-terser';
import path from 'path';

/*
 Path to the library output
 */
const PACKAGE_DIR = 'dist/first';
const PKG_JSON_PATH = `${PACKAGE_DIR}/package.json`;

/*
 List here all the modules you want to bundle
 Every other import will remain excluded
*/
const MODULES_TO_BUNDLE = [
	'is-glob',
        'is-extglob'
]

// Resolve the filepath to the fesm2015 file
const pkgJson = require(`./${PKG_JSON_PATH}`);
const fesmPath = path.join(PACKAGE_DIR, pkgJson.fesm2015);
const bundledFesmPath = path.join(PACKAGE_DIR, 'bundledFesm2015', path.basename(pkgJson.fesm2015));

/*
	This config bundles external modules together with a given bundle.
	In this example the FESM2015 bundle is used.
	The output is written to <PACKAGE_DIR>/bundledFesm2015/
 */

export default {
	input: fesmPath,
	inlineDynamicImports: false,
	preserveSymlinks: true,
	treeshake: false,
	output: {
		file: bundledFesmPath,
		format: 'es',
		sourcemap: true
	},
	external: id => !MODULES_TO_BUNDLE.some(mId => id.includes(mId)),
	plugins: [
		nodeResolve(),
		commonJs(),
		sourceMaps(),
		terser({
			ecma: 2018,
			format: {
				comments: false
			}
		})
	]
}
```

Steps: 
- create a js file, e.g. `rollup.config.js`, and put that code in
- adjust
  - `PACKAGE_DIR`: This is the path to your **compiled** library (created by `ng-packagr`)
  - `MODULES_TO_BUNDLE`: Here, list all the dependencies you want to bundle with your angular library
- (Optionally adjust everything else according to your needs)
- **run** rollup by entering `./node_modules/.bin/rollup -c rollup.config.js` 

### Results

I created a sample library where I imported `is-glob`, which I wanted to bundle together.
 `ng-packagr` is creating this as expected:
![image](https://user-images.githubusercontent.com/5240450/106738763-a21d2780-6618-11eb-8e1c-c9e717822641.png)

After running config with the config above, it looks like this:
![image](https://user-images.githubusercontent.com/5240450/106739235-32f40300-6619-11eb-8fcd-49df6ac0f8e6.png)
(The `is-glob has been bundled into the code`)

#### Pitfalls
If you external dependencies itself also has dependencies, you have to provide them in the config, too. **all of them**.
In this case `is-glob` was importing `is-extglob`.

Hope this helps 

cheers
flash ⚡
